src/2024h2/sandboxed-build-script.md

```diff
@@ -0,0 +1,214 @@
+that is not realistic and still requires non-negligible efforts on code reviews,
+especially when compared to the speed at which new versions are coming out.
+
+Besides the security and trust issues,
+in an unsandboxed build script,
+random network or file system access may occur and fail.
+These kinds of "side effects" are notoriously non-deterministic,
+and usually cause retries and rebuilds in build pipelines.
+Because the build is not deterministic,
+reproducibility cannot be easily achieved,
+making programs harder to trace and debug.
+
+There is one 2024 GSoC project ["Sandboxed and Deterministic Proc Macro using Wasm"][GSoC]
+experimenting with the possibility of using WebAssembly to sandbox procedural macros.
+While build scripts and proc-macros are different concepts at different levels,
+they share the same flaw — arbitrary code execution.
+Given that we already have experiments on the proc-macros side,
+it's better we start some groundwork on build scripts in parallel,
+and discuss the potential common interface for Cargo to configure them.
+
+[cargo-vet]: https://crates.io/crates/cargo-vet
+[cargo-crev]: https://crates.io/crates/cargo-crev
+[GSoC]: https://summerofcode.withgoogle.com/programs/2024/projects/kXG0mZoj
+
+### The next 6 months
+
+* Look at prior art in this domain, especially for potential blockers and challenges.
+* Prototype on sandboxing build scripts.
+  Currently looking at [WebAssembly System Interface (WASI)](WASI) and [Cackle].
+* Provide a way to opt-in sandboxed build scripts for Cargo packages,
+  and design a configurable interface to grant permissions to each crate.
+* Based on the results of those experiments,
+  consider whether the implementation should be a third-party Cargo plugin first,
+  or make it into Cargo as an unstable feature (with a proper RFC).
+
+[wasi]: https://wasi.dev/
+[Cackle]: https://github.com/cackle-rs/cackle
+
+### The "shiny future" we are working towards
+
+These could become future goals if this one succeeds:
+
+* The sandboxed build script feature will be opted-in at first when stabilized.
+  By the next Edition, sandboxed build scripts will be on by default,
+  hardening the supply chain security.
+* Cargo users only need to learn one interface for both sandboxed proc-macros and build scripts.
+  The configuration for build scripts will also cover the needs for sandboxed proc-macros,
+* Crates.io and the [`cargo info`] command display the permission requirements of a crate,
+  helping developers choose packages based on different security level needs.
+* The runtime of the sandbox environment is swappable,
+  enabling the potential support of remote execution without waiting for a first-party solution.
+  It also opens a door to hermetic builds.
+* If we choose something like WebAssembly as the runtime of the sandbox environment,
+  the build script binaries could be pre-built and downloadable from user-defined pre-built binary servers.
```

Comment:
This has all of the same problems as other pre-built binary strategies and I see it as unlikely.  Instead, this can offer more reusability with a remote shared cache.

Reply:
Removed. I don't think we have any plan to put this in the "shiny future" at this moment.